util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+    hex_format_str = '0' + str(bytes_per_word_ecc * 2) + 'x'
+    memory_words = '// OTP memory hexfile with {} x {}bit layout\n'.format(
+        num_words, bytes_per_word_ecc * 8)
+    log.info('Memory layout is {} x {}bit (with ECC)'.format(
+        num_words, bytes_per_word_ecc * 8))
+
+    for k in range(num_words):
+        # Assemble native OTP word and uniquify annotation for comments
+        word = 0
+        word_ann = OrderedDict()
+        for j in range(bytes_per_word):
+            idx = k * bytes_per_word + j
+            word += data[idx] << (j * 8)
+            if annotation[idx] not in word_ann:
+                word_ann.update({annotation[idx]: "1"})
+
+        # This prints the byte offset of the corresponding
+        # payload data in the memory map (excluding ECC bits)
+        word_ann = ' // {:06x}: '.format(k * bytes_per_word) + ', '.join(
+            word_ann.keys())
+
+        # ECC encode
+        word_bin = format(word, bin_format_str)
+        word_bin = ecc_encode(config, word_bin)
+        word_hex = format(int(word_bin, 2), hex_format_str)
+        memory_words += word_hex + word_ann + '\n'
+
+    log.info('Done.')
+
+    return memory_words
+
+
+class OtpMemImg(OtpMemMap):
+
+    # LC state object
+    lc_state = []
+
+    def __init__(self, lc_state_config, otp_mmap_config, img_config):
+
+        # Initialize memory map
+        super().__init__(otp_mmap_config)
+
+        # Initialize the LC state and OTP memory map objects first, since
+        # validation and image generation depends on them
+        self.lc_state = LcStEnc(lc_state_config)
+
+        # Validate memory image configuration
+        log.info('')
+        log.info('Parse OTP image specification.')
+
+        # Encryption sanity test with known test vector
+        enc_test = _present_64bit_encrypt(
+            int("0123456789abcdef", 16),
+            int("0123456789abcdef0123456789abcdef", 16))
+
+        if enc_test == int('0e9d28685e671dd6', 16):
+            log.info('Encryption module test OK')
+        else:
+            raise RuntimeError('Encryption module test failed')
+        log.info('')
+
+        if 'seed' not in img_config:
+            raise RuntimeError('Missing seed in configuration.')
+
+        img_config['seed'] = check_int(img_config['seed'])
+        log.info('Seed: {0:x}'.format(img_config['seed']))
+        log.info('')
+
+        # Re-initialize with seed to make results reproducible.
+        random.seed(OTP_IMG_SEED_DIVERSIFIER + int(img_config['seed']))
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Adding values to {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+        log.info('')
+        log.info('Parsing OTP image successfully completed.')
```

Comment:
I'd consider a check here to make sure that `img_config` doesn't have any keys that you didn't expect. A neat way to do it is to take a shallow copy of `img_config` at the start of the method and then delete keys as you use them. Then if `img_config` is non-empty when you get to this point, you know something was bad and you also have a list of the invalid keys.

Reply:
Yes that's a good idea.
I've done something along these lines, although I had to take the shallow copy later on during the methods as some of them use setdefault to set keys that may have been omitted.